cpp/src/arrow/ipc/writer.cc

```diff
@@ -263,6 +263,14 @@ class RecordBatchSerializer {
     out_->body_length = offset - buffer_start_offset_;
     DCHECK(bit_util::IsMultipleOf8(out_->body_length));
 
+    // copy given record batch's schema metadata to the serializer for serialization
+    auto const &metadata = batch.schema()->metadata();
```

Comment:
```suggestion
    const auto& metadata = batch.schema()->metadata();
```
for consistency

Reply:
I'm not sure about this change. If the overall schema has metadata, then won't this copy the schema metadata into every batch, duplicating it over and over?

Reply:
Maybe we should instead add overloads that also allow explicitly passing in metadata to include with the message. Thoughts @pitrou @westonpace?

Reply:
I think I agree.  This would make it too easy to write an inefficient file that contains a lot of duplicate metadata.  Adding an optional metadata field to the `WriteTable` method makes sense.

Reply:
I think there are two cases.

1) in one case, a user simply re-uses the overall schema that happens to have some metadata, and it is less likely this user wants to duplicate the schema metadata in all batches and discard them silently is okay (and duplicating it all record batches may be undesirable and problematic):
```python
    schema = pa.schema(
        [
            ("values", pa.int64()),
        ],
        metadata={"foo": "bar"},
    )
    writer = pa.RecordBatchFileWriter(
        ipc_file, schema
    )
    for i in range(num_batches):
        batch = pa.record_batch(
            [int_array],
            schema=schema # <=== re-use the overall schema having metadata
        )
        writer.write_batch(batch)
```

2) in the other case, a user provides metadata explicitly when creating a record batch, and discarding the metadata silently may not be desirable in this case:
```python
   schema = pa.schema(
        [
            ("values", pa.int64()),
        ],
        metadata={"foo": "bar"},
    )
    writer = pa.RecordBatchFileWriter(
        ipc_file, schema
    )
    for i in range(num_batches):
        batch = pa.record_batch(
            [int_array],
            names=["values"],
            metadata={"batch_id": str(i)}, # <=== pass a metadata explicitly
        )
        writer.write_batch(batch)
```
The underlying implementation is difficult to tell these two cases.

Some current API, like `pyarrow.record_batch`, already allows users to specify metadata explicitly. If we provide an overloaded function asking users to provide metadata, it seems we make the `metadata` in these APIs useless (or only useful when the record batch is in memory). The code may look like:
```python
batch = pa.record_batch(
    [int_array],
    names=["values"] # <== `metadata` passed here will never be used by `write_batch` so users won't pass this parameter any more
)
writer.write_batch(batch, metadata={"batch_id": str(i)})
```

Maybe a boolean flag indicating if the metadata in record batch should be serialized?
```python
batch = pa.record_batch(
    [int_array],
    names=["values"],
    metadata={"batch_id": str(i)}
)
writer.write_batch(batch, serializing_metadata=True)
```

Reply:
A boolean parameter seems like a fine idea to me.  Let's call it `serialize_metadata` to keep the tenses consistent.

Reply:
@westonpace 
>  Let's call it serialize_metadata to keep the tenses consistent
Sure. I am not a native English speaker, and feel free to let me know if there is better naming for this parameter.

@westonpace @lidavidm 
I will give it a try but I am not completely sure what kind of API compatibility is required. Could you give me some suggestions on what the overloaded function signatures are desirable?

Currently in the C++ API, we have:
```C++
class ARROW_EXPORT RecordBatchWriter {
 public:
  virtual ~RecordBatchWriter();

  /// \brief Write a record batch to the stream
  ///
  /// \param[in] batch the record batch to write to the stream
  /// \return Status
  virtual Status WriteRecordBatch(const RecordBatch& batch) = 0;

  /// \brief Write possibly-chunked table by creating sequence of record batches
  /// \param[in] table table to write
  /// \return Status
  Status WriteTable(const Table& table);

  /// \brief Write Table with a particular chunksize
  /// \param[in] table table to write
  /// \param[in] max_chunksize maximum length of table chunks. To indicate
  /// that no maximum should be enforced, pass -1.
  /// \return Status
  virtual Status WriteTable(const Table& table, int64_t max_chunksize);
```

Do we need to keep all existing APIs unchanged or is it allowed to introduce some compatible API changes?
Some options I can think of:
1) option 1, keep all existing APIs unchanged and add overloaded APIs only
```C++
virtual Status WriteRecordBatch(const RecordBatch& batch) = 0;
virtual Status WriteRecordBatch(const RecordBatch& batch, bool serialize_metadata); // new overload API
Status WriteTable(const Table& table);
virtual Status WriteTable(const Table& table, int64_t max_chunksize);
virtual Status WriteTable(const Table& table, int64_t max_chunksize, bool serialize_metadata); // new overload API
```

2) option 2, add a default parameter to existing APIs
```C++
virtual Status WriteRecordBatch(const RecordBatch& batch, bool serialize_metadatda=false) = 0; // this will require modification to all sub classes
Status WriteTable(const Table& table);
virtual Status WriteTable(const Table& table, int64_t max_chunksize, bool serialize_metadata=false);
```

I assume we expect option 1, is it correct?

Additionally, there may be some higher level binding libraries (Python/Ruby/etc) using these APIs, and they may have to be changed to take advantage of these new APIs, what is our strategy for managing them for such a change? Do I need to change them all in this PR (I think I can try changing Python at least but I am not sure if I am familiar with all the binding languages to make such a change) or should I separate the changes into several PRs? Thanks.

Reply:
For the bindings: we can tackle them separately. CI will catch us if we make any breaking changes that need to be handled here. It's only Python/R/GLib (Ruby) we have to worry about (Java/Go have a few bindings I think but not at this level). 

For the APIs: I would prefer adding overloads. CC @pitrou for opinions too.